sbc_compassion/models/correspondence.py

```diff
@@ -614,33 +618,38 @@ def get_chars(t):
     @api.model
     def process_commkit(self, commkit_data):
         """ Update or Create the letter with given values. """
-        letter_mapping = mapping.new_onramp_mapping(self._name, self.env)
-        letter_ids = list()
-        process_letters = self
-        for commkit in commkit_data.get('Responses', [commkit_data]):
-            vals = letter_mapping.get_vals_from_connect(commkit)
-            published_state = 'Published to Global Partner'
-            is_published = vals.get('state') == published_state
-
-            # Write/update letter
-            kit_identifier = vals.get('kit_identifier')
-            letter = self.search([('kit_identifier', '=', kit_identifier)])
-            if letter:
-                # Avoid to publish twice a same letter
-                is_published = is_published and letter.state != published_state
-                letter.write(vals)
-            else:
-                if 'id' in vals:
-                    del vals['id']
-                letter = self.with_context(no_comm_kit=True).create(vals)
-
-            if is_published:
-                process_letters += letter
+        self.process_lock.acquire()
+        try:
+            letter_mapping = mapping.new_onramp_mapping(self._name, self.env)
+            letter_ids = list()
+            process_letters = self
+            for commkit in commkit_data.get('Responses', [commkit_data]):
+                vals = letter_mapping.get_vals_from_connect(commkit)
+                published_state = 'Published to Global Partner'
+                is_published = vals.get('state') == published_state
+
+                # Write/update letter
+                kit_identifier = vals.get('kit_identifier')
+                letter = self.search([('kit_identifier', '=', kit_identifier)])
+                if letter:
+                    # Avoid to publish twice a same letter
+                    is_published = is_published and \
+                                   letter.state == published_state
```

Comment:
Please keep the previous definition of `is_published`, otherwise you break the code at line 645 which will not test the same case anymore. `is_published = is_published and letter.state != published_state`. 
Adapt your test at line 638 with `if is_published or letter.state != published_state:`